R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+#' @param skip.markers whether to skip exporting markers
+#' @param skip.expr.matrix whether to skip exporting expression matrix
+#' @param skip.metadata whether to skip exporting metadata
+#' @param skip.reductions whether to skip exporting reductions
+#' @param port on which port to run UCSC cellbrowser webserver after export
+#' @param ... specifies the metadata fields to export. To supply a field and its
+#' human readable name, pass name as \code{field="name"} parameter.
+#'
+#' @return This function exports Seurat object as a set of tsv files
+#' to \code{dir} directory, copying the \code{markers.file} if it is
+#' passed. It also creates the default \code{cellbrowser.conf} in the
+#' directory. This directory could be read by \code{cbBuild} to
+#' create a static website viewer for the dataset. If \code{cb.dir}
+#' parameter is passed, the function runs \code{cbBuild} (if it is
+#' installed) to create this static website in \code{cb.dir} directory.
+#' If \code{port} parameter is passed, it also runs the webserver for
+#' that directory and opens a browser.
+#'
+#' @author Maximilian Haeussler, Nikolay Markov
+#'
+#' @importFrom tools file_ext
+#' @importFrom utils browseURL
+#' @importFrom reticulate py_module_available import
+#' @importFrom Seurat Project Idents GetAssayData Embeddings FetchData
+#'
+#' @export
+#'
+#' @name CellBrowser
+#' @rdname CellBrowser
+#'
+#' @importFrom methods slot
+#' @importFrom utils packageVersion
+#' @importFrom reticulate py_module_available import
+#'
+#' @examples
+#' \dontrun{
+#' ExportToCellbrowser(pbmc_small, dataset.name = "PBMC", dir = "out")
+#' }
+#'
+ExportToCellbrowser <- function(
+  object,
+  dir,
+  dataset.name = Project(object = object),
+  reductions = NULL,
+  markers.file = NULL,
+  cluster.field = NULL,
+  cb.dir = NULL,
+  port = NULL,
+  use.mtx = FALSE,
+  meta.fields = NULL,
+  meta.fields.names = NULL,
+  matrix.slot = "counts",
+  markers.n = 100,
+  skip.markers = FALSE,
+  skip.expr.matrix = FALSE,
+  skip.metadata = FALSE,
+  skip.reductions = FALSE
+) {
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("This script requires that Seurat (V2 or V3) is installed")
+  }
+  message("Seurat Version installed: ", packageVersion(pkg = "Seurat"))
+  object.version <- as.character(x = slot(object = object, name = 'version'))
+  message("Object was created with Seurat version ", object.version)
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
```

Comment:
Could a lot of this be done with math on `numeric_version` objects? Also, SeuratWrappers only works on Seurat v3.0.1 or higher; you cannot have SeuratWrappers and Seurat v2 installed at the same time

Reply:
1) What stops me from installing seuratWrappers when I have Seurat2 installed? 

2) nice! I rewrote this with "numeric_version" now.

Reply:
[This does](https://github.com/satijalab/seurat-wrappers/blob/master/DESCRIPTION#L36). We explicitly require Seurat v3.0.1 or higher with SeuratWrappers. If you try to install SeuratWrappers with Seurat v2.X installed, R will automatically install the latest CRAN version of Seurat (or GitHub version, which is synced with the CRAN version, if you installed from GitHub) and overwrite your Seurat v2.X installation. If you then go back and install Seurat v2.X, I don't believe SeuratWrappers will load as we import a lot of Seurat v3.X-specific code.

Reply:
OK, Thanks for the explanation. I'd still really prefer to keep this particular function backwards-compatible, as it's just a few lines and this allows us to use this particular function across the three different Seurat versions. The version problems with Seurat are getting worse and worse and this function isolates us from the Seurat versions.